### PR TITLE
fix: prevent file traversal using helm file values param and application details api

### PR DESCRIPTION
**Pull request by alexmt**
Signed-off-by: Alexander Matyushentsev <AMatyushentsev@gmail.com>

PR introduces missing checks in repo server API that prevents file traversal attack:

* adds missing checks of helm value files in `repository.GetAppDetails` API
* adds missing checks of helm `fileParameters` in `repository.GenerateManifest` API

### CODE REVIEW

File: util/helm/resolved.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		absWorkDir, err := filepath.Abs(appPath)
+		if err != nil {
+			return "", false, resolveFailure(repoRoot, err)
+		}
+		path = filepath.Join(absWorkDir, path)
+	} else {
+		path = filepath.Join(absRepoPath, path)
+	}
+
+	// Ensure any symbolic link is resolved before we
```
Comment (crenshaw-dev) [inline util/helm/resolved.go]: Incomplete comment

File: reposerver/repository/repository.go [hunk 2]
```diff
@@ -766,7 +615,7 @@ func helmTemplate(appPath string, repoRoot string, env *v1alpha1.Env, q *apiclie
 				return nil, err
 			}
 			defer file.Close()
-			templateOpts.Values = append(templateOpts.Values, p)
+			templateOpts.Values = append(templateOpts.Values, helm.ResolvedFilePath(p))
```
Comment (crenshaw-dev) [inline reposerver/repository/repository.go]: Won't this always be out-of-path? 

> If dir is the empty string, CreateTemp uses the default directory for temporary files

https://pkg.go.dev/os#CreateTemp

We could write the values file into the repo directory temporarily. If we're allowing concurrent use of the repo, there's a chance someone could snatch someone else's values file, so we should probabluy upgrade from `ioutil.TempFile` (which in 1.17 is the same as `os.CreateTemp`) to something containing a UUIDv4.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Quick first pass. Will come back after coffee.  :-)

File: util/helm/resolved.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
+package helm
```
Comment (jannfis) [inline util/helm/resolved.go]: Maybe this should go into a dedicated, generic package? E.g. `util/file` or `util/path`?

Comment (alexmt) [reply util/helm/resolved.go]: Agree. Moved it to `util/io/path` . Our `util` has so many top-level packages so I wanted to start grouping them.

File: reposerver/repository/repository.go (hunk 2 above)
Comment (alexmt) [reply reposerver/repository/repository.go]: Agree about using UUIDv4 - updated.

We want to make sure attacker won't access out-of-path file. In this case, path is generated with no user input so safe to use out-of-path location.

**Comment by alexmt** [comment]
After moving the `resolved.go` to separate package I've noticed that jsonnet also might be vulnerable:

https://github.com/argoproj/argo-cd/blob/e2a2881a429b780e0a9ef134f6fccc0f4a7a80c6/reposerver/repository/repository.go#L1237


It is not vulnerable right now, because go-jsonnet does not support symlinks, but support was [considered](https://github.com/google/go-jsonnet/pull/443#discussion_r466376905) and might be added in a future. So added a check to jsonnet as well.

**Review by crenshaw-dev — APPROVED** [review]
LGTM!

Should there be any docs changes to indicate the new path limitations for jsonnet/Helm in 2.3? Or will release notes be enough?

**Review by jannfis — APPROVED** [review]
LGTM

**Comment by alexmt** [comment]
Closing this PR and creating PR in upstream: https://github.com/argoproj/argo-cd/pull/8606

**Pull request closed without merging**